hack/generator/pkg/astmodel/resource.go

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"go/ast"
+	"go/token"
+)
+
+// CreateResourceDefinitions creates definitions for a resource
+func CreateResourceDefinitions(name *TypeName, specType *StructType, statusType *StructType, idFactory IdentifierFactory) (TypeDefiner, []TypeDefiner) {
+
+	var others []TypeDefiner
+
+	var specName *TypeName
+	if specType != nil {
+		specName = NewTypeName(name.PackageReference, name.Name()+"Spec")
+		specDef, specOthers := specType.CreateDefinitions(specName, idFactory)
+
+		others = append(append(others, specDef), specOthers...)
+	} else {
+		panic("spec must always be provided")
+	}
+
+	var statusName *TypeName
+	if statusType != nil {
+		statusName = NewTypeName(name.PackageReference, name.Name()+"Status")
+		statusDef, statusOthers := statusType.CreateDefinitions(statusName, idFactory)
+
+		others = append(append(others, statusDef), statusOthers...)
+	}
+
+	this := &ResourceDefinition{typeName: name, spec: specName, status: statusName, isStorageVersion: false}
+
+	return this, others
+}
+
+// ResourceDefinition represents an ARM resource
+type ResourceDefinition struct {
+	typeName         *TypeName
+	spec             *TypeName
+	status           *TypeName
+	isStorageVersion bool
+	description      *string
+}
+
+// assert that ResourceDefinition implements TypeDefiner
+var _ TypeDefiner = &ResourceDefinition{}
+
+func (definition *ResourceDefinition) Name() *TypeName {
+	return definition.typeName
+}
+
+func (definition *ResourceDefinition) Type() Type {
+	return definition.spec // TODO?????
```

Comment:
Not an issue for now since we don’t yet set the `status` field but I will be doing so later.

One fix for this is that we remove `Type()` from `TypeDefiner` and instead lift up the two functions we need (`References(TypeName)` and `RequiredImports` into the `TypeDefiner` interface). These are then straightforward to implement by multiplexing over `spec`/`status` here.

Reply:
Removing `Type()` from `TypeDefiner` _sounds_ like it reduces the number of abstractions we need to keep track of at once when working on the code, and therefore might be a good idea. #possiblymistaken

Reply:
Yes, I won’t do it in this PR though.

Reply:
Expanded comment to include this info.